java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java

```diff
@@ -58,13 +58,16 @@ TypeSubstitution getTypeSubstitution(JavaSymbol.MethodJavaSymbol method, JavaTyp
     return substitution;
   }
 
-  JavaType getReturnType(@Nullable JavaType returnType, JavaType site, boolean parametrizedMethodCall, TypeSubstitution substitution) {
+  JavaType getReturnType(@Nullable JavaType returnType, JavaType defSite, JavaType callSite, boolean parametrizedMethodCall, TypeSubstitution substitution) {
     if (returnType == null) {
       // case of constructors
       return returnType;
     }
-    JavaType resultType = applySiteSubstitution(returnType, site);
-    if (isRawTypeOfParametrizedType(site) && !parametrizedMethodCall) {
+    JavaType resultType = applySiteSubstitution(returnType, defSite);
+    if(callSite != defSite) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)